File: cplusplus/numberOfIslands.cpp
#include "globals.h"

void dfs(int r, int c, int maxRow, int maxCol, vector<vector<char>>& grid, vector<vector<bool>>& visited) {

	// Check if position is safe
	if(r < 0 || c < 0 || r >= maxRow || c >= maxCol || grid[r][c] == '0' || visited[r][c]) {
		return;
	}
	visited[r][c] = true;

	// NOTE: Can be expanded to include diagonals if desired
	vector<int> rowNeighbors = {0,  0, -1, 1};
	vector<int> colNeighbors = {1, -1,  0, 0};

	for(int i = 0; i < rowNeighbors.size(); i++) {
		dfs(r+rowNeighbors[i], c+colNeighbors[i], maxRow, maxCol, grid, visited);
	}
}

int numIslands(vector<vector<char>>& grid) {

	const int maxRow = grid.size();
	const int maxCol = grid[0].size();

	// SIZE first, then INSERTION of actual element
	vector<vector<bool>> visited(maxRow, vector<bool>(maxCol, false));

	// Perform DFS for each element in the grid
	int count = 0;
	for(int r = 0; r < maxRow; r++) {
		for(int c = 0; c < maxCol; c++) {
			if(grid[r][c] == '1' && !visited[r][c]) {
				dfs(r, c, maxRow, maxCol, grid, visited);
				count++;
			}
		}
	}
	return count;
}

int main() {

	vector<vector<char>> grid = {{'1', '1', '0', '0', '0'},
								 {'1', '1', '0', '0', '0'},
								 {'0', '0', '1', '0', '0'},
								 {'0', '0', '0', '1', '1'}};

	int ans = numIslands(grid);
	cout << "Expected Answer: 3" << endl;
	cout << "Real Answer: " << ans << endl;
}

File: python/numberOfIslands.py
class Solution:
    def dfs(self, r, c, max_rows, max_cols, grid, visited):
        if r < 0 or c < 0 or r >= max_rows or c >= max_cols or grid[r][c] == "0" or visited[r][c]:
            return

        visited[r][c] = True
        row_neighbors = [1, 0, 0, -1]
        col_neighbors = [0, 1, -1, 0]

        for i in range(4):
            new_r = r + row_neighbors[i]
            new_c = c + col_neighbors[i]
            self.dfs(new_r, new_c, max_rows, max_cols, grid, visited)

    def numIslands(self, grid: List[List[str]]) -> int:
        max_rows = len(grid)
        max_cols = len(grid[0])

        res = 0
        visited = [[False for _ in range(max_cols)] for _ in range(max_rows)]

        for r in range(max_rows):
            for c in range(max_cols):
                if grid[r][c] == "1" and not visited[r][c]:
                    res += 1
                    self.dfs(r, c, max_rows, max_cols, grid, visited)
        return res
